python_modules/dagster/dagster/core/definitions/time_window_partitions.py

```diff
@@ -34,6 +34,7 @@ class TimeWindowPartitionsDefinition(
             ("timezone", str),
             ("fmt", str),
             ("end_offset", int),
+            ("start_day_of_window", Optional[int]),
```

Comment:
I'm not happy with this name or the others i've chosen in this file. Going to spend some time thinking or alternatives, and I'm open to suggestions

Reply:
Some options:
* split_day
*

Reply:
https://github.com/dagster-io/dagster/issues/6057 requests that similar functionality be added to the hourly and daily partitions too. so the name should be broad enough to encompass all these use cases. `offset` isn't too bad - maybe each implementation of `TimeWindowPartitionDefinition` could have a custom version of `offset` (ie `daily_partitioned_config` would have `hour_offset` and `monthly_partitioned_config` would have `day_offset`)

Reply:
So would we use `day_of_week_offset` in this case? I think that naming convention sounds pretty solid. A couple things to make sure jive:
* We have an `end_offset` param.
* I think we'd likely want to support multiple offset arguments? E.g. `monthly_partitioned_config(day_offset=3, hour_offset=2)` for partitions that start at 2 am on the 3rd day of every month.

Reply:
What about adding:
`minute_offset`
`hour_offset`
`day_offset` - corresponds to day of week for weekly partition and day of month for monthly partition

We'd expose each offset param in the partitions where it makes sense.

Reply:
I've been working through implementing this and running into some complexities...

Right now, all time window partitions except hourly format the partition keys as `Y-m-d` but the addition of hour and minute offsets means we need to switch to `Y-m-d-h-m` format. seems like nbd, but the time window partitions expect the `start` to be formatted in the same way as the partition keys. so that means every partition with a `start` in format `Y-m-d` (which is what all of our testing and docs use) no longer works. 

One option is that we could make it so time window partitions can handle a `start` that's a different format than what the keys will be.

But this has made me think even more about the [design question](https://github.com/dagster-io/dagster/pull/7125#pullrequestreview-914691148) i had in this PR. I'm starting to lean in the direction that we should support users passing the start in `Y-m-d-h-m` format and base the windows off of that. 

Example:

```
@daily_partitioned_config(start="2022-2-18-02:15")
# creates partitions (2022-2-18-02:15, 2022-2-19-02:15), (2022-2-19-02:15, 2022-2-20-02:15) ...

@weekly_partitioned_config(start="2022-2-18-03:10")
# creates partitions (2022-2-18-03:10, 2022-2-25-03:10), (2022-2-25-03:10, 2022-3-04-03:10)...

@daily_partitioned_config(start="2022-2-16", fmt="%Y-%m-%d")
# creates partitions (2022-2-16, 2022-2-17), (2022-2-17, 2022-2-18)...
```
This reads cleaner to me and someone reading the partition will know exactly when the first partition starts, rather than needing to know implementations details like that a weekly partition actually begins the sunday after the start unless otherwise specified. 

It also reduces collisions if a user specifies a datetime as start. What is the expected behavior in a situation like this?
`@daily_partitioned_config(start=datetime(2022, 2, 18, 2, 15), minute_offset=30, hour_offset=5)`


To reduce the number of breaking changes, we could make the default string format for `start` `Y-m-d` and document how to change the format to support hour/minute/day offsets. But this would still introduce breaking changes for weekly and monthly partitions because the start date of the partition would change to match the use specified time

Reply:
Deleted this comment initially - adding it back for posterity though. I ran into a couple sections of code that I thought would break the approach i propose and wanted to take some time to understand the current implementation before considering an api change in addition

Original message:

I've been working through implementing this and running into some complexities...

Right now, all time window partitions except hourly format the partition keys as `Y-m-d` but the addition of hour and minute offsets means we need to switch to `Y-m-d-h-m` format. seems like nbd, but the time window partitions expect the `start` to be formatted in the same way as the partition keys. so that means every partition with a `start` in format `Y-m-d` (which is what all of our testing and docs use) no longer works. 

One option is that we could make it so time window partitions can handle a `start` that's a different format than what the keys will be.

But this has made me think even more about the [design question](https://github.com/dagster-io/dagster/pull/7125#pullrequestreview-914691148) i had in this PR. I'm starting to lean in the direction that we should support users passing the start in `Y-m-d-h-m` format and base the windows off of that. 

Example:

```
@daily_partitioned_config(start="2022-2-18-02:15")
# creates partitions (2022-2-18-02:15, 2022-2-19-02:15), (2022-2-19-02:15, 2022-2-20-02:15) ...

@weekly_partitioned_config(start="2022-2-18-03:10")
# creates partitions (2022-2-18-03:10, 2022-2-25-03:10), (2022-2-25-03:10, 2022-3-04-03:10)...

@daily_partitioned_config(start="2022-2-16", fmt="%Y-%m-%d")
# creates partitions (2022-2-16, 2022-2-17), (2022-2-17, 2022-2-18)...
```
This reads cleaner to me and someone reading the partition will know exactly when the first partition starts, rather than needing to know implementations details like that a weekly partition actually begins the sunday after the start unless otherwise specified. 

It also reduces collisions if a user specifies a datetime as start. What is the expected behavior in a situation like this?
`@daily_partitioned_config(start=datetime(2022, 2, 18, 2, 15), minute_offset=30, hour_offset=5)`


To reduce the number of breaking changes, we could make the default string format for `start` `Y-m-d` and document how to change the format to support hour/minute/day offsets. But this would still introduce breaking changes for weekly and monthly partitions because the start date of the partition would change to match the use specified time

Reply:
ok code confusion is resolved - implemented the following api 
```
@hourly_partitioned_config(start, minute_offset)
@daily_partitioned_config(start, minute_offset, hour_offset)
@weekly_partitioned_config(start, minute_offset, hour_offset, day_offset)
@monthly_partitioned_config(start, minute_offset, hour_offset, day_offset)
```
where `minute_offset` `hour_offset` and `day_offset` are all optional 

My current plan is to add docs explaining how partitions are determined based on the `start` and offsets so that it's well documented that weekly partitions start on sunday unless otherwise specified, etc.